phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRepairRegionScanner.java

```diff
@@ -0,0 +1,426 @@
+                }
+                del.addDeleteMarker(cell);
+            }
+        }
+        List<Mutation> indexMutations = prepareIndexMutationsForRebuild(indexMaintainer, put, del);
+        Collections.reverse(indexMutations);
+        for (Mutation mutation : indexMutations) {
+            byte[] indexRowKey = mutation.getRow();
+            List<Mutation> mutationList = expectedIndexMutationMap.get(indexRowKey);
+            if (mutationList == null) {
+                mutationList = new ArrayList<>();
+                mutationList.add(mutation);
+                expectedIndexMutationMap.put(indexRowKey, mutationList);
+            } else {
+                mutationList.add(mutation);
+            }
+        }
+    }
+
+    private void repairIndexRows(Map<byte[], List<Mutation>> indexMutationMap,
+                                 List<Mutation> indexRowsToBeDeleted,
+                                 IndexToolVerificationResult verificationResult) throws IOException {
+        try {
+            int batchSize = 0;
+            List<Mutation> indexUpdates = new ArrayList<Mutation>(maxBatchSize);
+            for (List<Mutation> mutationList : indexMutationMap.values()) {
+                indexUpdates.addAll(mutationList);
+                batchSize += mutationList.size();
+                if (batchSize >= maxBatchSize) {
+                    ungroupedAggregateRegionObserver.checkForRegionClosing();
+                    region.batchMutate(indexUpdates.toArray(new Mutation[indexUpdates.size()]),
+                            HConstants.NO_NONCE, HConstants.NO_NONCE);
+                    batchSize = 0;
+                    indexUpdates = new ArrayList<Mutation>(maxBatchSize);
+                }
+            }
+            if (batchSize > 0) {
+                ungroupedAggregateRegionObserver.checkForRegionClosing();
+                region.batchMutate(indexUpdates.toArray(new Mutation[indexUpdates.size()]),
+                        HConstants.NO_NONCE, HConstants.NO_NONCE);
+            }
+            batchSize = 0;
+            indexUpdates = new ArrayList<Mutation>(maxBatchSize);
+            for (Mutation mutation : indexRowsToBeDeleted) {
+                indexUpdates.add(mutation);
+                batchSize ++;
+                if (batchSize >= maxBatchSize) {
+                    ungroupedAggregateRegionObserver.checkForRegionClosing();
+                    region.batchMutate(indexUpdates.toArray(new Mutation[indexUpdates.size()]),
+                            HConstants.NO_NONCE, HConstants.NO_NONCE);
+                    batchSize = 0;
+                    indexUpdates = new ArrayList<Mutation>(maxBatchSize);
+                }
+            }
+            if (batchSize > 0) {
+                ungroupedAggregateRegionObserver.checkForRegionClosing();
+                region.batchMutate(indexUpdates.toArray(new Mutation[indexUpdates.size()]),
+                        HConstants.NO_NONCE, HConstants.NO_NONCE);
```

Comment:
I feel the same logic of checking if the region is closed and then sending the mutations is duplicated in multiple places and we can move it to a function to reduce the code duplication.

Reply:
Question: Here we are calling `region.batchMutate` to update the index table but in IndexRebuildRegionScanner we call `indexHTable.batch()`. Why the difference ?

Reply:
I will reduce the code duplication here. When the index table is local (in the case of repair) we use the region API and when the index table is remote (in the case of rebuild), we use the table API.

Reply:
Done